pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -61,172 +103,188 @@ type ecrTokenGetter struct {
+		if err != nil {
+			klog.Errorf("while decoding token for endpoint %v %v", data.ProxyEndpoint, err)
+			return cfg
+		}
+		parts := strings.SplitN(string(decodedToken), ":", 2)
+		user := parts[0]
+		password := parts[1]
+		creds := credentialprovider.DockerConfigEntry{
+			Username: user,
+			Password: password,
+			// ECR doesn't care and Docker is about to obsolete it
+			Email: "not@val.id",
+		}
+		entry := cacheEntry{
+			expiresAt:   *data.ExpiresAt,
+			credentials: creds,
+			host:        parsed.host,
+		}
+		if err := p.cache.Add(entry); err != nil {
+			klog.Errorf("while adding entry to cache %v", err)
+			return cfg
+		}
+		cfg[entry.host] = entry.credentials
+	}
 	return cfg
 }
 
-func newEcrProvider(region string, getter tokenGetter) *ecrProvider {
+/*PROVIDER*/
+// ecrProvider is a DockerConfigProvider that gets and refreshes tokens
+// from AWS to access ECR.
+type ecrProvider struct {
+	registryURLTemplate string
+	cache               cache.Store
+	getter              tokenGetter
+}
+
+var _ credentialprovider.DockerConfigProvider = &ecrProvider{}
+
+func newECRProvider(template string, getter tokenGetter) *ecrProvider {
 	return &ecrProvider{
-		region:    region,
-		regionURL: registryURL(region),
-		getter:    getter,
+		registryURLTemplate: template,
+		// cache:               cache.NewTTLStore(stringKeyFunc, 1*time.Hour),
+		cache:  cache.NewStoreWithPolicy(stringKeyFunc, &ecrExpirationPolicy{}),
+		getter: getter,
 	}
 }
 
+// init registers a credential provider for the specified region.
+// It creates a provider for each registryURLTemplate, in order to
+// support cross-region ECR access.
+func init() {
+	credentialprovider.RegisterCredentialProvider("aws-ecr-partition-standard",
+		newECRProvider(registryURLTemplateStandard, nil))
+
+	credentialprovider.RegisterCredentialProvider("aws-ecr-partition-china",
+		newECRProvider(registryURLTemplateChina, nil))
+}
+
 // Enabled implements DockerConfigProvider.Enabled for the AWS token-based implementation.
 // For now, it gets activated only if AWS was chosen as the cloud provider.
 // TODO: figure how to enable it manually for deployments that are not on AWS but still
 // use ECR somehow?
 func (p *ecrProvider) Enabled() bool {
-	if p.region == "" {
-		klog.Errorf("Called ecrProvider.Enabled() with no region set")
-		return false
-	}
-
-	getter := &ecrTokenGetter{svc: ecr.New(session.New(&aws.Config{
-		Credentials: nil,
-		Region:      &p.region,
-	}))}
-	getter.svc.Handlers.Build.PushFrontNamed(request.NamedHandler{
-		Name: "k8s/user-agent",
-		Fn:   request.MakeAddToUserAgentHandler("kubernetes", version.Get().String()),
-	})
-	getter.svc.Handlers.Sign.PushFrontNamed(request.NamedHandler{
-		Name: "k8s/logger",
-		Fn:   awsHandlerLogger,
-	})
-	p.getter = getter
-
 	return true
 }
 
-// LazyProvide implements DockerConfigProvider.LazyProvide. Should never be called.
-func (p *ecrProvider) LazyProvide() *credentialprovider.DockerConfigEntry {
+// LazyProvide is lazy
```

Comment:
:grin: Might want to explain why this is unused.

Reply:
If LazyProvide doesn't do anything here are we keeping it to avoid having to change the other credential providers?

Reply:
We were thinking it would be better as a separate PR to remove and change it everywhere so it doesn't make this PR super chunky.

Reply:
That makes sense, when we open the PR let's mention that and maybe at a TODO comment to say we will remove this shortly after